### ISSUE TITLE
Enhancement: [strict-void-return] False-positive report on promisified exec

### ISSUE THREAD

**Issue by RareScrap** [typescript-eslint/typescript-eslint#12015]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/strict-void-return/

### Description

It's not possible to use this rule with promisified stuff

### Fail

```TypeScript
var replace = 'me';
```

### Pass

```TypeScript
import { exec } from 'node:child_process';
import { promisify } from 'node:util';

const execPromisified = promisify(exec)
```

### Additional Info

The issue is requested by @bradzacher in [typescript-eslint Discord discussion](https://discord.com/channels/1026804805894672454/1465587382395867147/1465673242306940970)

**Comment by kirkwaiblinger** [comment]
[Playground link](https://typescript-eslint.io/play/#ts=5.9.3&showAST=es&fileType=.tsx&code=JYWwDg9gTgLgBAbzgUwB7IMZwL5wGZQQhwDkAdhACbIBcGAFsADaUD6YhGyAztyQNwAoUJFiI4HIsG7A8ATxz5CxclVoBXGMwGDBGCGW7w0mAArLps4MkpwAvBIsz5AChMYAlEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugNwHtKAE1rRE%2BWNCbooAdwCGkyODABfECqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by kirkwaiblinger** [comment]
I'm not sure, but this might be happening because `exec()` both returns a value _and_ takes a node `(err, value) =>` callback. If I understand correctly, `promisify()` ignores the return value of a function and returns a promise which is either fulfilled with `value` or rejected with `err`. Therefore, it may typed to accept a `void`-returning function. (I tried to validate this but the types are really complex so I haven't dived deep enough into this to know.)

---

Here's my attempt at a simplified repro (but as said this could be incorrect):

[Playground](https://typescript-eslint.io/play/#ts=5.9.3&showAST=es&fileType=.tsx&code=C4TwDgpgBAcg9gEwgYQIYBt0CNUGMDWAPACoB8UAvFABQQBOdAXFAK4B2%2BbcA7mwDRQAbhhYRmxAJSVyguAEsEAKEUAzdrmBy4bKAGc5AWzDoIABTpwDc-SpAlS1Fc2q4szeEjSYcBe1Ioy8ghSAN6KUFB0EMAsdDrU-uRsENxQ5pbWEPbU4RE0Ubpw6IIQAlEAVhAaiVBheXkq1LQMAsLoojV19XlyKjT0dFIVVcDNgwDcud1QEOi60AVFJdRtHZPTAL4SU1BbihvKSLjoqFFQamwaWjqgYHK4GB4QAGLqmtoubrCIKBjYeER2JweGxSBJmLIFJMjiczhcrtooFwkK9Lu82AB1OTAAAWACVorE2J93D8vP9fECuLwwcxdMA6HI2ABzSaKfRGEzpKw2EDUW73R4-VEItgScZQAD0kqRcCgYAsWBMBnZhmMZgsPN6fORLze1yxuIJMTi4qlMtQun0zLYTOZUAABvTGSyHdJHZCEA6AHRAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugNwHtKAE1rRE%2BWNCbooAdwCGkyODABfECqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false)

### Code

```ts
type NodeCallback<T> = (err: unknown, value: T) => void

function simplePromisify<T>(f: (cb: NodeCallback<T>) => void) {
  return () => new Promise<T>(
    (resolve, reject) => {
      f((err, value) => {
        if (err) reject(err);
        else resolve(value);
      })
    })
}

declare function typicalNodeFunction(cb: NodeCallback<unknown>): void;
declare function nodeFunctionWithReturn(cb: NodeCallback<unknown>): string;

simplePromisify(typicalNodeFunction); // no problem
simplePromisify(nodeFunctionWithReturn); // assigning `string` => `void`.
```

### Eslint config

```json
{
  "rules": {
    "@typescript-eslint/strict-void-return": "warn"
  }
}
```

### TypeScript config

```json
{
  "compilerOptions": {
    "strict": true
  }
}
```

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `8.54.0` |
| `@typescript-eslint/parser` | `8.54.0` |
| `TypeScript` | `5.9.3` |
| `ESLint` | `9.39.2` |
| `node` | `web` |

---

Another piece of evidence for this idea is that I don't get this report when doing `promisify(stat)`, which is `void`-returning.

More investigation needed though.

**Comment by bradzacher** [comment]
This might be a bug in `@types/node` then?
Perhaps the type should be `=> unknown` instead of `=> void`.
Cos promisfy just ignores the return type.

**Comment by kirkwaiblinger** [comment]
Heh, we'll see. Gives me flashbacks to https://github.com/typescript-eslint/typescript-eslint/issues/10959,  https://github.com/microsoft/TypeScript/issues/44980

**Comment by darshjme-codes** [comment]
**Triage note:**

False-positive report from `strict-void-return` on promisified `exec` callback.

**Issue:** Rule doesn't recognize that `util.promisify(exec)` properly handles the callback return type transformation.

**Suggested enhancement:** Add configuration option to allow specific promisified functions or improve type analysis for `util.promisify` patterns.

**Labels applied:** `enhancement: plugin rule option`, `triage` ✓  
**Discussion:** 4 comments (community engagement good)

**Comment by JoshuaKGoldberg** [comment]
Going back over this, it looks to me like this really is an upstream bug in `@types/node`. Kirk's simplified repro seems correct. Does anybody want to protest this issue being closed here? @bradzacher @kirkwaiblinger @RareScrap

**Comment by bradzacher** [comment]
It's a painful thing in the ecosystem where so many people abuse `void` because TS makes it work.

The types should probably be `unknown` yeah - but that might be a hard change to land upstream now. But I think that somebody pushing for the ecosystem to be "more correct" would be good here.

**Comment by JoshuaKGoldberg** [comment]
Closing out as it's been >1 month. We are pretty confident this is a bug in DefinitelyTyped -> `@types/node`, in that `promisify` should type its function parameter as `=> unknown` rather than `=> void`. Cheers!

Co-authored-by: @StyleShit
